### PR TITLE
[Android] Add message UI for speedreader

### PR DESCRIPTION
**Pull request by wchen342**
Resolves brave/brave-browser/issues/28769

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. On Android, go to `brave://flags` and enable speedreader.
2. Open a tab and go to a supported url, for example https://www.theguardian.com/technology/2023/feb/25/nokia-launches-diy-repairable-budget-android-phone.
3. The page shall be displayed in speedreader.
4. Now tap on "View original", the original web page shall be displayed instead.
5. Go to another tab, and than switch back. A pop-up shall be displayed under url bar asking to enable speedreader.
6. Tap "Enable" shall put the page into speedreader mode again.

### CODE REVIEW

**Comment by iefremov** [comment]
can you link a spec to the ticket please?

File: browser/ui/android/strings/android_brave_strings.grd
Comment (SergeyZhukovsky) [inline browser/ui/android/strings/android_brave_strings.grd:3555]: maybe it's better to have `View` as it is done in upstream for a simplified page view? cc @rmcfadden3 for the words above. @wchen342 could you please post a picture here with the dialog, so Ryan has a context where does it appear?

**Review by iefremov — APPROVED** [review]
browser/speedreader and chromium_src lgtm

**Comment by iefremov** [comment]
@boocmp FYI

File: browser/speedreader/speedreader_tab_helper.cc
Comment (SergeyZhukovsky) [inline browser/speedreader/speedreader_tab_helper.cc:124]: let's add a Preference to control that as well, place it near the simplified mode view (Settings->Appearance). Call it same way as we have it on desktop. Let's make it hidden if flag is off and it is off by default. When a speedreader selected let's hide the simplified page view control and only show it when speedreader is disabled.

File: android/java/org/chromium/chrome/browser/speedreader/BraveSpeedReaderManager.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/speedreader/BraveSpeedReaderManager.java:192]: we need to add a pref check here as well

File: browser/ui/android/strings/android_brave_strings.grd
Comment (rmcfadden3) [reply browser/ui/android/strings/android_brave_strings.grd:3555]: Thanks, @SergeyZhukovsky — yeah, a screenshot would help so I can understand full context. But, at a glance I'd say "View page with Speedreader?" makes intuitive sense.

Also note we've capitalized the word "Speedreader" elsewhere, so I've adjusted that above.

Comment (wchen342) [reply browser/ui/android/strings/android_brave_strings.grd:3555]: @rmcfadden3 Here is a screenshot of the popup:

Screenshot

**Comment by wchen342** [comment]
@iefremov There is no spec for the Android UI at this time.

File: browser/speedreader/speedreader_tab_helper.cc
Comment (wchen342) [reply browser/speedreader/speedreader_tab_helper.cc]: I have added the preference and also hide the simplified view setting.

File: android/java/org/chromium/chrome/browser/settings/AppearancePreferences.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/settings/AppearancePreferences.java:196]: do we really need to relaunch to enable it?

**Review by samartnik — APPROVED** [review]
Java patching looks good

File: android/java/org/chromium/chrome/browser/speedreader/BraveSpeedReaderManager.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/speedreader/BraveSpeedReaderManager.java:125]: why can't we use it while in a desktop view mode?

File: android/java/org/chromium/chrome/browser/settings/AppearancePreferences.java
Comment (wchen342) [reply android/java/org/chromium/chrome/browser/settings/AppearancePreferences.java:196]: Yeah I've tested it, without relaunching it doesn't change.

File: android/java/org/chromium/chrome/browser/speedreader/BraveSpeedReaderManager.java
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/speedreader/BraveSpeedReaderManager.java:125]: let's remove it and check what happens? Maybe there are complications with an upstream simplified view in that view mode? But we should be ok with our Speedreader implementation.

File: browser/ui/android/strings/android_brave_strings.grd
Comment (SergeyZhukovsky) [inline browser/ui/android/strings/android_brave_strings.grd:3562]: Could you please post the screenshot of the above preference as well so @rmcfadden3 knows the context as well

Comment (wchen342) [reply browser/ui/android/strings/android_brave_strings.grd:3562]: Ok here is a screenshot for the preference. I took the text from Desktop:

Screenshot

File: browser/ui/android/strings/android_brave_strings.grd
Comment (rmcfadden3) [reply browser/ui/android/strings/android_brave_strings.grd:3555]: Thanks, @wchen342 . Yeah, I'd say `View page with Spreedreader` works well here. And note the capitalization on "Speedreader"

Comment (wchen342) [reply browser/ui/android/strings/android_brave_strings.grd:3555]: Yes I have made the "S" capitalized. Somehow it wasn't reflected in the conversation page I think.

**Review by SergeyZhukovsky — APPROVED** [review]
++. @rmcfadden3 could you please approve if words look good?

**Comment by rmcfadden3** [comment]
> ++. @rmcfadden3 could you please approve if words look good?

Thanks, @SergeyZhukovsky — yeah, we should adjust the Settings copy for consistency. Most other menu items begin with a verb, and none of the other menu descriptions include a subjective value proposition like "…saving you time"

Let's go with:

Header: `Enable Speedreader`
Body: `Automatically loads articles in reader mode`

Also, does having Speedreader enabled automatically load **_all_** articles in reader mode? I thought it was only certain articles, but not all. If that's the case, we could add a modifier to the body copy, using something like `Automatically loads eligible articles in reader mode`

File: browser/ui/android/strings/android_brave_strings.grd
Comment (mkarolin) [inline browser/ui/android/strings/android_brave_strings.grd:3558]: @rmcfadden3 Is `Speedreader` a translatable term?

**Comment by wchen342** [comment]
@rmcfadden3 I have updated the wording.

File: browser/ui/android/strings/android_brave_strings.grd
Comment (rmcfadden3) [reply browser/ui/android/strings/android_brave_strings.grd]: @mkarolin — I believe the answer is yes: We translate basically everything except for the word "Brave" itself, so we would also want to translate the word "speedreader" here.

But CC @laurentiugruia to confirm. He should have final say on that.

Comment (laurentiugruia) [reply browser/ui/android/strings/android_brave_strings.grd]: Thank you for the tag, @rmcfadden3. 

Generally yes but we have exceptions sometimes, depending on the feature name and if translating it to the target language is tricky and/or affects the intended message.

In this case I've looked around in Transifex for quite a few languages and almost all translators kept it as "Speedreader", guessing due to the above reasoning.

I'm fine then with tagging it as "notranslate".

Comment (wchen342) [reply browser/ui/android/strings/android_brave_strings.grd]: On Android we can only make the whole message untranslatable, not just a word. In that case which one shall I mark non-traslatable? @rmcfadden3

Comment (mkarolin) [reply browser/ui/android/strings/android_brave_strings.grd]: I would suggest just marking `IDS_ENABLE_SPEEDREADER_TITLE` as `translateable="false"`. This way we won't upload it for translations.

Comment (wchen342) [reply browser/ui/android/strings/android_brave_strings.grd]: Updated.

**Review by rmcfadden3 — APPROVED** [review]
Copy changes LGTM!

**Pull request merged**
